src/dev/lupio/lupio_rng.cc

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_rng.hh"
+
+#include <random>
+
+#include "debug/LupioRNG.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioRNG.hh"
+
+/* CTRL fields */
+#define LUPIO_RNG_IRQE  0x1
+/* STAT fields */
+#define LUPIO_RNG_BUSY  0x1
+
+namespace gem5
+{
+
+using namespace RiscvISA;
+
+LupioRNG::LupioRNG(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    byteOrder(params.byte_order),
+    plic(params.plic),
+    int_id(params.lupio_rng_int_id),
+    user_seed(params.seed)
+{
+    mt.seed(user_seed);
+    DPRINTF(LupioRNG, "LupioRNG initalized with seed: %d\n", user_seed);
+}
+
+uint64_t
+LupioRNG::lupioRNGRead(uint8_t addr)
+{
+    uint32_t r = 0;
+    switch (addr >> 2) {
+        // returns a random number
+        case LUPIO_RNG_RAND: {
```

Comment:
I don't think you need curly brackets for any of the cases here (nor in this read function or the write function below).

Curly brackets are used to define a block, in case a new local variable needs to be declared. AFAIK, you can declare new local variables anywhere with C++, so it seems unnecessary.